src/twisted/web/_websocket_impl.py

```diff
@@ -0,0 +1,457 @@
+        """
+
+    def detachProducer(self) -> None:
+        """
+        Detach a previously attached L{IPushProducer} from this transport.
+        """
+
+
+@dataclass
+class ConnectionRejected(Exception):
+    """
+    A websocket connection was rejected by an HTTP response.
+
+    @ivar response: The HTTP response that describes the rejection.
+    """
+
+    response: Response
+
+
+class WebSocketProtocol(TypingProtocol):
+    """
+    An object that conforms to L{WebSocketProtocol} can receive all the events
+    from a websocket connection.
+    """
+
+    def negotiationStarted(self, transport: WebSocketTransport) -> None:
+        """
+        An underlying transport (e.g.: a TCP connection) has been established;
+        negotiation of the websocket transport has begun.
+        """
+
+    def negotiationFinished(self) -> None:
+        """
+        Negotiation is complete: a bidirectional websocket channel is now fully
+        established.
+        """
+
+    def pongReceived(self, payload: bytes) -> None:
+        """
+        A 'pong' message was received.
+        """
+
+    def textMessageReceived(self, message: str) -> None:
+        """
+        A text message was received from the peer.
+        """
+
+    def bytesMessageReceived(self, data: bytes) -> None:
+        """
+        A bytes message was received from the peer.
+        """
+
+    def connectionLost(self, reason: Failure) -> None:
+        """
+        The websocket connection was lost.
+        """
+
+
+_WSP = TypeVar("_WSP", covariant=True, bound=WebSocketProtocol)
+_Bootstrap = Callable[[Union[WSConnection, Connection], ITransport], None]
+_log = Logger()
+
+
+class WebSocketServerFactory(TypingProtocol[_WSP]):
+    """
+    A L{WebSocketServerFactory} is a factory for a particular kind of
+    L{WebSocketProtocol} that implements server-side websocket listeners via
+    L{WebSocketResource}.
+    """
+
+    def buildProtocol(self, request: Request) -> _WSP:
+        """
+        To conform to L{WebSocketServerFactory}, you must implement a
+        C{buildProtocol} method which takes a L{Request
+        <twisted.web.server.Request>} and returns a L{WebSocketProtocol}.
+
+        @return: a L{WebSocketProtocol} that will handle the inbound
+            connection.
+        """
+
+
+class WebSocketClientFactory(TypingProtocol[_WSP]):
+    """
+    A L{WebSocketClientFactory} is a factory for a particular kind of
+    L{WebSocketProtocol} that implements client-side websocket listeners via
+    L{WebSocketClientEndpoint}.
+    """
+
+    def buildProtocol(self, url: str) -> _WSP:
+        """
+        To conform to L{WebSocketServerFactory}, you must implement a
+        C{buildProtocol} method which takes a string representing an URL and
+        returns a L{WebSocketProtocol}.
+
+        @return: a L{WebSocketProtocol} that will handle the outgoing
+            connection.
+        """
+
+
+@dataclass(frozen=True)
+class WebSocketClientEndpoint:
```

Comment:
It still doesn't make sense to me why we need to use

`endpoint = WebSocketClientEndpoint.new(reactor, "ws://localhost:8080/webskt.rpy")`

and not just

`endpoint = WebSocketClientEndpoint(reactor, "ws://localhost:8080/webskt.rpy")`

----------

Why not setup the delegation as part of init ?

I agree not to trigger side-effects as part of `init`.

------

Maybe I am missing something.

I know that the goal is to stop using zope.interface

I was expecting to see the equivalent protocol for `IStreamClientEndpoint` define via the stdlib [Protocol](https://typing.python.org/en/latest/spec/protocol.html)

Reply:
I wrote a whole blog post about this https://blog.glyph.im/2025/04/stop-writing-init-methods.html :)